compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/FragmentFunctionGenerator.kt

```diff
@@ -0,0 +1,81 @@
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.IrBody
+import org.jetbrains.kotlin.ir.util.declareSimpleFunctionWithOverrides
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi.KtBlockCodeFragment
+
+
+class FragmentFunctionGenerator(
+    declarationGenerator: FragmentDeclarationGenerator,
+    private val fragmentInfo: EvaluatorFragmentInfo
+) : FragmentDeclarationGeneratorExtension(declarationGenerator) {
+
+    fun generateFunctionForFragment(ktFile: KtBlockCodeFragment): IrSimpleFunction =
+        declareFragmentFunction(
+            IrDeclarationOrigin.DEFINED,
+        ) {
+            generateExpressionBody(ktFile.getContentElement())
+        }
+
+    private fun declareFragmentFunction(
+        origin: IrDeclarationOrigin,
+        generateBody: BodyGenerator.() -> IrBody
+    ): IrSimpleFunction =
+        declareFragmentFunctionInner(fragmentInfo.methodDescriptor, origin).buildWithScope { irFunction ->
+            generateFragmentFunctionParameterDeclarationsAndReturnType(irFunction)
+            irFunction.body = createBodyGenerator(irFunction.symbol).generateBody()
+        }
+
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
+    private fun declareFragmentFunctionInner(descriptor: FunctionDescriptor, origin: IrDeclarationOrigin): IrSimpleFunction =
+        context.symbolTable.declareSimpleFunctionWithOverrides(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            origin,
+            descriptor
+        )
+
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
+    private fun generateFragmentFunctionParameterDeclarationsAndReturnType(
+        irFunction: IrSimpleFunction
+    ) {
+        declarationGenerator.generateScopedTypeParameterDeclarations(irFunction)
+        irFunction.returnType = irFunction.descriptor.returnType!!.toIrType()
+        generateFragmentValueParameterDeclarations(irFunction)
+    }
+
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
+    private fun generateFragmentValueParameterDeclarations(
+        irFunction: IrSimpleFunction
+    ) {
+        val functionDescriptor = irFunction.descriptor
+        functionDescriptor.valueParameters.forEachIndexed { index, valueParameterDescriptor ->
+            irFunction.valueParameters += declareParameter(valueParameterDescriptor).apply {
+                context.capturedDescriptorToFragmentParameterMap[fragmentInfo.parameters[index]] = this.symbol
+            }
+        }
+    }
+
+    private fun declareParameter(descriptor: ValueParameterDescriptor, name: Name? = null) =
+        context.symbolTable.declareValueParameter(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            IrDeclarationOrigin.DEFINED,
+            descriptor, descriptor.type.toIrType(),
+            (descriptor as? ValueParameterDescriptor)?.varargElementType?.toIrType(),
+            name,
+            isAssignable = true
```

Comment:
Could you clarify why it should be assignable? Maybe it deserves a comment with an explanation...

Reply:
It's needed for the approach that the evaluator takes to modifications to captured variables. Captures are passed as parameters to the fragment code, modifications modify the _parameters_ directly, and the evaluator then reads back the new value from the stack frame of the interpreter/VM running the fragment, which is then written back to the original captured variable in the JVM being debugged.

Reply:
Thanks, got it. It would be great to add such explanation as a comment.

Reply:
Done.